File: net/JetBrains.Serialization/src/Dataholders/CertificateInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.X509;

namespace JetBrains.Serialization;

[JsonObject(MemberSerialization.OptIn)]
public class CertificateInfo : IEncodableInfo
{
  [JsonProperty("XCertificateInfo")] private XCertificateInfo _xCertificateInfo;

  [JsonProperty("SignatureAlgorithm")] private AlgorithmInfo _signatureAlgorithm;

  [JsonProperty("SignatureData")] private TextualInfo _signatureData;

  [JsonConstructor]
  private CertificateInfo(XCertificateInfo xCertificateInfo, AlgorithmInfo signatureAlgorithm,
    TextualInfo signatureData)
  {
    _xCertificateInfo = xCertificateInfo;
    _signatureAlgorithm = signatureAlgorithm;
    _signatureData = signatureData;
  }

  public static CertificateInfo GetInstance(X509CertificateStructure x509CertificateStructure)
    => new CertificateInfo(
      XCertificateInfo.GetInstance(x509CertificateStructure.TbsCertificate),
      new AlgorithmInfo(x509CertificateStructure.SignatureAlgorithm),
      TextualInfo.GetInstance(new DerBitString(x509CertificateStructure.GetSignatureOctets()))
    );

  public static CertificateInfo GetInstance(AttributeCertificate attributeCertificate)
    => new CertificateInfo(
      XCertificateInfo.GetInstance(attributeCertificate),
      new AlgorithmInfo(attributeCertificate.SignatureAlgorithm),
      TextualInfo.GetInstance(new DerBitString(attributeCertificate.GetSignatureOctets()))
    );

  public static IEncodableInfo GetInstance(Asn1Object obj)
  {
    switch (obj)
    {
      case DerSequence sequence:
        return GetInstance(X509CertificateStructure.GetInstance(sequence));

      case Asn1TaggedObject taggedObject:
        return new TaggedObjectInfo(
          taggedObject.IsExplicit(),
          taggedObject.TagNo,
          GetInstance(AttributeCertificate.GetInstance(taggedObject.GetObject()))
        );

      default:
        throw new ArgumentException("Unexpected object type");
    }
  }

  private DerSequence ToDLSequence() =>
    new List<IEncodableInfo?>
    {
      _xCertificateInfo,
      _signatureAlgorithm,
      _signatureData
    }.ToPrimitiveDerSequence();

  public Asn1Encodable ToPrimitive() => ToDLSequence();
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/CertificateInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.toDLSequence
import com.jetbrains.signatureverifier.serialization.toDLSet
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.*
import org.bouncycastle.asn1.x509.AttributeCertificate
import org.bouncycastle.asn1.x509.Certificate
import org.bouncycastle.cert.X509AttributeCertificateHolder
import org.bouncycastle.cert.X509CertificateHolder
import org.bouncycastle.util.Store

@Serializable
data class CertificateInfo(
  val xCertificateInfo: XCertificateInfo,
  val signatureAlgorithm: AlgorithmInfo,
  val signatureData: TextualInfo
) : EncodableInfo {
  companion object {
    fun getInstance(certificateHolder: X509CertificateHolder) = CertificateInfo(
      XCertificateInfo.getInstance(certificateHolder),
      AlgorithmInfo(certificateHolder.signatureAlgorithm),
      TextualInfo.getInstance(DERBitString(certificateHolder.signature))
    )

    fun getInstance(certificateHolder: X509AttributeCertificateHolder) = CertificateInfo(
      XCertificateInfo.getInstance(certificateHolder),
      AlgorithmInfo(certificateHolder.signatureAlgorithm),
      TextualInfo.getInstance(DERBitString(certificateHolder.signature))
    )

    fun getInstance(obj: ASN1Object): EncodableInfo = when (obj) {
      is DLSequence -> {
        getInstance(
          X509CertificateHolder(
            Certificate.getInstance(
              obj
            )
          )
        )
      }

      is DLTaggedObject -> {
        TaggedObjectInfo(
          obj.isExplicit,
          obj.tagNo,
          getInstance(
            X509AttributeCertificateHolder(
              AttributeCertificate.getInstance(
                obj.baseObject
              )
            )
          )
        )
      }

      else -> throw IllegalArgumentException("Unexpected object type")
    }
  }

  private fun toDlSequence(): DLSequence =
    listOf(
      xCertificateInfo.toPrimitive(),
      signatureAlgorithm.toPrimitive(),
      signatureData.toPrimitive()
    ).toDLSequence()

  fun toX509CertificateHolder() = X509CertificateHolder(
    Certificate.getInstance(toPrimitive())
  )

  override fun toPrimitive(): ASN1Primitive =
    toDlSequence().toASN1Primitive()
}

fun recreateCertificatesFromStore(store: Store<X509CertificateHolder>): ASN1Set =
  store.getMatches(null).toList().map { it.toASN1Structure() }.toDLSet()